src/open_company/resources/company.clj

```diff
@@ -138,44 +140,103 @@
-                            (sections-for)) ; add/replace the :sections property
-          section-names (flatten (vals (:sections interim-company)))
-          final-company (->> interim-company
-                          (author-for author section-names) ; add/replace the :author in the sections
-                          (updated-for timestamp section-names))] ; add/replace the :updated-at in the sections
-      (if (true? (valid-company final-company))
-        (do
-          ;; create the each section in the DB
-          (doseq [section-name section-names]
-            (let [section (-> (get final-company (keyword section-name))
-                              (assoc :company-slug (:slug final-company))
-                              (assoc :section-name section-name))]
-              (common/create-resource common/section-table-name section timestamp)))
-          ;; create the company in the DB
-          (common/create-resource table-name final-company timestamp))
-        false))))
+(defn placeholder-sections [company-slug]
+  (reduce (fn [s sec]
+            (assoc s
+                   (-> sec :name keyword)
+                   (-> sec
+                       (assoc :section-name (-> sec :name keyword))
+                       (assoc :company-slug company-slug)
+                       (assoc :placeholder true)
+                       (dissoc :name))))
+          {}
+          (filter :core common/sections)))
+
+(defn add-placeholder-sections [company]
+  (-> (placeholder-sections (:slug company))
+      (merge company)
+      (sections-for)))
+
+;; ----- Create saveable company doc ----
+
+(defn real-sections [company]
+  (med/remove-vals :placeholder (select-keys company common/section-names)))
+
+(defn complete-real-sections
+  [company user]
+  (let [rs (real-sections company)
+        add-info (fn [[k section-data]]
+                   [k (-> section-data
+                          (assoc :author (common/author-for-user user))
+                          (assoc :company-slug (:slug company))
+                          (assoc :section-name k))])]
+    (merge company (into {} (map add-info rs)))))
+
+(s/defn ->company :- common/Company
+  [company-props user]
+  (let [slug (or (:slug company-props) (slug/find-available-slug (:name company-props) (taken-slugs)))]
+    (-> company-props
+        (assoc :slug slug)
+        (update :currency #(or % "USD"))
+        (assoc :org-id (:org-id user))
+        (complete-real-sections user)
+        (categories-for)
+        (sections-for))))
+
+(s/defn add-updated-at
+  [{:keys [section-name] :as section} :- common/Section ts]
+  (let [notes-section? (contains? common/notes-sections section-name)]
+    (cond-> (assoc section :updated-at ts)
+      (and notes-section? (:notes section)) (assoc-in [:notes :updated-at] ts))))
+
+(s/defn create-company!
+  [company :- common/Company]
+  (s/validate common/Company company) ; throw if invalid
+  (let [real-sections     (real-sections company)
```

Comment:
Don't line up the right side of these let assignments. We don't do it throughout the rest of the code base.